src/NLog/Internal/FilePathLayout.cs

```diff
@@ -138,56 +141,78 @@ public Layout GetLayout()
         #region Implementation of IRenderable
 
         /// <summary>
-        /// Render, as cleaned if requested.
+        /// Render the raw filename from Layout
         /// </summary>
         /// <param name="logEvent">The log event.</param>
         /// <returns>String representation of a layout.</returns>
-        private string GetCleanedFileName(LogEventInfo logEvent)
+        private string GetRenderedFileName(LogEventInfo logEvent)
         {
             if (cleanedFixedResult != null)
             {
                 return cleanedFixedResult;
             }
+
             if (_layout == null)
             {
                 return null;
             }
 
-            var result = _layout.Render(logEvent);
-            if (_cleanupInvalidChars)
-            {
-                return CleanupInvalidFilePath(result);
-            }
-            return result;
+            return _layout.Render(logEvent);
         }
 
-        public string Render(LogEventInfo logEvent)
+        /// <summary>
+        /// Convert the raw filename to a correct filename
+        /// </summary>
+        /// <param name="rawFileName">The filename generated by Layout.</param>
+        /// <returns>String representation of a correct filename.</returns>
+        private string GetCleanFileName(string rawFileName)
         {
-            var rendered = GetCleanedFileName(logEvent);
-            if (String.IsNullOrEmpty(rendered))
+            var cleanFileName = rawFileName;
+            if (_cleanupInvalidChars && cleanedFixedResult == null)
             {
-                return rendered;
+                cleanFileName = CleanupInvalidFilePath(rawFileName);
             }
 
             if (_filePathKind == FilePathKind.Absolute)
             {
-                return rendered;
+                return cleanFileName;
             }
+
 #if !SILVERLIGHT
             if (_filePathKind == FilePathKind.Relative)
             {
-                return Path.Combine(_baseDir, rendered);
                 //use basedir, faster than Path.GetFullPath
+                cleanFileName = Path.Combine(_baseDir, cleanFileName);
+                return cleanFileName;
             }
 #endif
             //unknown, use slow method
-            return Path.GetFullPath(rendered);
+            cleanFileName = Path.GetFullPath(cleanFileName);
+            return cleanFileName;
+        }
 
+        public string Render(LogEventInfo logEvent)
+        {
+            var rawFileName = GetRenderedFileName(logEvent);
+            if (string.IsNullOrEmpty(rawFileName))
+            {
+                return rawFileName;
+            }
+
+            if ((!_cleanupInvalidChars || cleanedFixedResult != null) && _filePathKind == FilePathKind.Absolute)
+                return rawFileName; // Skip clean filename string-allocation
+
+            if (string.CompareOrdinal(_cachedPrevRawFileName, rawFileName) == 0 && _cachedPrevCleanFileName != null)
```

Comment:
_cachedPrevCleanFileName will never be null, and when it is then _cachedPrevRawFileName is also null. I just have a secret agenda when it comes to the PR with PoolSetup.